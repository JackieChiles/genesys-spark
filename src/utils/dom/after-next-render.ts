export function afterNextRender(callback: FrameRequestCallback): void {
  requestAnimationFrame(() => requestAnimationFrame(callback));
}
