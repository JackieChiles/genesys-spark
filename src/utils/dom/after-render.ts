export function afterRender(callback: FrameRequestCallback): void {
  requestAnimationFrame(() => requestAnimationFrame(callback));
}
