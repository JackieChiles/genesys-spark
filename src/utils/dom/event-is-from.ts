/**
 * Checks to see if an event originated within an element matching a specific selector.
 */
export function eventIsFrom(selector: string, event: Event): boolean {
  // Don't search for matches above the node where we captured the event;
  const exitTarget = event.currentTarget as HTMLElement;
  // Start with the oringal event target
  let checkTarget = event.target as HTMLElement;
  while (checkTarget !== exitTarget && checkTarget !== null) {
    if (checkTarget.matches(selector)) {
      return true;
    } else {
      checkTarget = checkTarget.parentElement;
    }
  }
  return false;
}
