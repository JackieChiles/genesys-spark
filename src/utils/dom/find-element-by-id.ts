export function findElementById(
  root: HTMLElement,
  forElementId: string
): HTMLElement {
  let rootNode = root.getRootNode();
  let forElement: HTMLElement;

  while (rootNode && !forElement) {
    forElement = (rootNode as Document)?.getElementById(forElementId);
    rootNode = rootNode.getRootNode();
  }
  return forElement;
}
