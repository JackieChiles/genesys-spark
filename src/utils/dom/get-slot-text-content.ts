export function getSlotTextContent(
  root: HTMLElement,
  slotName: string
): string {
  return root.querySelector(`[slot=${slotName}]`)?.textContent;
}
