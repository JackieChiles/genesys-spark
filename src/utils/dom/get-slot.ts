export function getSlot(
  element: HTMLElement,
  slotName: string
): HTMLSlotElement {
  return element.querySelector(`[slot=${slotName}]`);
}
