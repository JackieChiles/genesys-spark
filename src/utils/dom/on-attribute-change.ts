export function onHiddenChange(
  element: HTMLElement,
  callback: (hidden: boolean) => void
): MutationObserver {
  const observer = new MutationObserver(mutations => {
    mutations.forEach(() => {
      callback(element.hidden);
    });
  });

  observer.observe(element, {
    attributes: true,
    attributeFilter: ['hidden']
  });

  return observer;
}

export function onDisabledChange(
  element: HTMLInputElement,
  callback: (disabled: boolean) => void
): MutationObserver {
  const observer = new MutationObserver(mutations => {
    mutations.forEach(mutation => {
      if (mutation.attributeName === 'disabled') {
        callback(element.disabled);
      }
    });
  });

  observer.observe(element, { attributes: true });

  return observer;
}
