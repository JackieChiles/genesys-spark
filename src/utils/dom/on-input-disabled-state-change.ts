import { calculateInputDisabledState } from './calculate-input-disabled-state';

export function onInputDisabledStateChange(
  element:
    | HTMLButtonElement
    | HTMLFieldSetElement
    | HTMLInputElement
    | HTMLOptGroupElement
    | HTMLOptionElement
    | HTMLSelectElement
    | HTMLTextAreaElement
    | HTMLGuxDropdownElement,
  callback: (disabled: boolean) => void
): MutationObserver {
  const fieldSet = element.closest('fieldset');

  const observer = new MutationObserver(mutations => {
    mutations.forEach(mutation => {
      if (mutation.attributeName === 'disabled') {
        callback(calculateInputDisabledState(element));
      }
    });
  });

  observer.observe(element, { attributes: true });

  if (fieldSet) {
    observer.observe(fieldSet, { attributes: true });
  }

  return observer;
}
