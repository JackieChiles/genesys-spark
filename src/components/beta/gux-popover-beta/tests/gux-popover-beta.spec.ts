import { newSpecPage } from '@stencil/core/testing';
import * as floatingUi from '@floating-ui/dom';
import MutationObserver from 'mutation-observer';
import { GuxPopoverBeta } from '../gux-popover-beta';

const components = [GuxPopoverBeta];
const language = 'en';
const html = `          
  <div>
    <div id="popover-target">
      Example Element
    </div>
    <gux-popover-beta id="popover-example" position="top" for="popover-target">
      <span slot="title">Title</span>
      <div>popover content</div>
    </gux-popover-beta>
  </div>`;

describe('gux-popover-beta', () => {
  beforeEach(async () => {
    global.MutationObserver = MutationObserver;
  });

  afterEach(async () => {
    jest.spyOn(floatingUi, 'autoUpdate').mockRestore();
  });

  describe('#render', () => {
    it('should render popover', async () => {
      jest.spyOn(floatingUi, 'autoUpdate').mockReturnValue(jest.fn());

      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxPopoverBeta);
      expect(page.root).toMatchSnapshot();
    });
  });
  describe('#interaction', () => {
    it('should open and close', async () => {
      const cleanupSpy = jest.fn();
      const autoUpdateSpy = jest
        .spyOn(floatingUi, 'autoUpdate')
        .mockReturnValue(cleanupSpy);

      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxPopoverBetaElement;
      const elementWrapper = element.shadowRoot.querySelector(
        '.gux-popover-wrapper'
      );
      expect(elementWrapper).not.toHaveClass('gux-hidden');
      expect(autoUpdateSpy).toBeCalledTimes(1);
      expect(cleanupSpy).not.toBeCalled();

      element.isOpen = false;
      await page.waitForChanges();
      expect(elementWrapper).toHaveClass('gux-hidden');
      expect(cleanupSpy).toBeCalledTimes(1);

      element.isOpen = true;
      await page.waitForChanges();
      expect(autoUpdateSpy).toBeCalledTimes(2);
      expect(cleanupSpy).toBeCalledTimes(1);
    });
  });
});
