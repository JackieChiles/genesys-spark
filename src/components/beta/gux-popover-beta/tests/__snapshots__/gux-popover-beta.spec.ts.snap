// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-popover-beta #render should render popover 1`] = `
<gux-popover-beta for="popover-target" id="popover-example" position="top">
  <mock:shadow-root>
    <div class="gux-popover-wrapper" data-placement>
      <div class="gux-arrow"></div>
      <div class="gux-popover-header">
        <slot name="title"></slot>
      </div>
      <div class="gux-popover-content">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <span slot="title">
    Title
  </span>
  <div>
    popover content
  </div>
</gux-popover-beta>
`;
